File: ui/src/models/agent.ts
export type AgentPhase =
	| "idle"
	| "investigating"
	| "plan-review"
	| "implementing"
	| "interrupted"
	| "code-review"
	| "fix-ready"
	| "ready-to-merge"
	| "completed";

export type AgentRunPhase = "investigation" | "implementation" | "fix" | "chat";
export type AgentRunStatus = "running" | "succeeded" | "failed" | "cancelled" | "interrupted";
export type ReviewStage = "plan" | "implementation";
export type AgentAction =
	| "start-investigation"
	| "approve-plan"
	| "request-plan-changes"
	| "approve-implementation"
	| "request-implementation-changes"
	| "start-fix"
	| "create-worktree"
	| "start-agent"
	| "commit-worktree"
	| "merge-worktree"
	| "complete-without-merge"
	| "resume"
	| "cancel";

export interface AgentWorkflow {
	projectId: string;
	taskId: string;
	phase: AgentPhase;
	activeRunId?: string;
	ompSessionId?: string;
	codexSessionId?: string;
	chatSessionId?: string;
	worktreePath?: string;
	worktreeBranch?: string;
	worktreeCommit?: string;
	resumePhase?: AgentRunPhase;
	updatedAt: string;
}

export interface AgentRun {
	id: string;
	projectId: string;
	taskId: string;
	phase: AgentRunPhase;
	status: AgentRunStatus;
	codexThreadId?: string;
	ompSessionId?: string;
	codexSessionId?: string;
	finishedAt?: string;
	exitCode?: number;
	summary?: string;
	tests?: string[];
	error?: string;
	logPath?: string;
}

export interface ReviewComment {
	id: string;
	projectId: string;
	taskId: string;
	stage: ReviewStage;
	body: string;
	createdAt: string;
	runId?: string;
}

export interface AgentTaskSnapshot {
	workflow: AgentWorkflow;
	chatSessionId?: string;
	runs: AgentRun[];
	reviewComments: ReviewComment[];
	dirtyFiles: string[];
	diff?: string;
	adapterState: "running" | "stopped";
	interrupted: boolean;
}

export interface CodexStatus {
	installed: boolean;
	loggedIn: boolean;
	version?: string;
	error?: string;
	installCommand?: string;
	loginCommand: string;
	docsUrl: string;
}

export type AgentStatus = CodexStatus;
export type OMPStatus = CodexStatus;

export interface AgentEvent {
	type?: string;
	projectId: string;
	taskId: string;
	runId?: string;
	message?: string;
	taskChanged?: boolean;
}

File: internal/models/agent.go
package models

import "time"

type AgentPhase string
type AgentRunPhase string
type AgentRunStatus string
type ReviewStage string

const (
	AgentPhaseIdle          AgentPhase = "idle"
	AgentPhaseInvestigating AgentPhase = "investigating"
	AgentPhasePlanReview    AgentPhase = "plan-review"
	AgentPhaseImplementing  AgentPhase = "implementing"
	AgentPhaseInterrupted   AgentPhase = "interrupted"
	AgentPhaseCodeReview    AgentPhase = "code-review"
	AgentPhaseFixReady      AgentPhase = "fix-ready"
	AgentPhaseCompleted     AgentPhase = "completed"

	AgentRunPhaseInvestigation  AgentRunPhase = "investigation"
	AgentRunPhaseImplementation AgentRunPhase = "implementation"
	AgentRunPhaseFix            AgentRunPhase = "fix"
	AgentRunPhaseChat           AgentRunPhase = "chat"

	AgentRunStatusRunning     AgentRunStatus = "running"
	AgentRunStatusSucceeded   AgentRunStatus = "succeeded"
	AgentRunStatusFailed      AgentRunStatus = "failed"
	AgentRunStatusCancelled   AgentRunStatus = "cancelled"
	AgentRunStatusInterrupted AgentRunStatus = "interrupted"

	ReviewStagePlan           ReviewStage = "plan"
	ReviewStageImplementation ReviewStage = "implementation"
)

type AgentWorkflow struct {
	ProjectID      string        `json:"projectId"`
	TaskID         string        `json:"taskId"`
	Phase          AgentPhase    `json:"phase"`
	ActiveRunID    string        `json:"activeRunId,omitempty"`
	CodexSessionID string        `json:"codexSessionId,omitempty"`
	ChatSessionID  string        `json:"chatSessionId,omitempty"`
	ResumePhase    AgentRunPhase `json:"resumePhase,omitempty"`
	UpdatedAt      time.Time     `json:"updatedAt"`
}

type AgentRun struct {
	ID             string         `json:"id"`
	ProjectID      string         `json:"projectId"`
	TaskID         string         `json:"taskId"`
	Phase          AgentRunPhase  `json:"phase"`
	Status         AgentRunStatus `json:"status"`
	CodexThreadID  string         `json:"codexThreadId,omitempty"`
	CodexSessionID string         `json:"codexSessionId,omitempty"`
	StartedAt      time.Time      `json:"startedAt"`
	FinishedAt     *time.Time     `json:"finishedAt,omitempty"`
	ExitCode       *int           `json:"exitCode,omitempty"`
	Summary        string         `json:"summary,omitempty"`
	Tests          []string       `json:"tests,omitempty"`
	Error          string         `json:"error,omitempty"`
	LogPath        string         `json:"logPath"`
}

type ReviewComment struct {
	ID        string      `json:"id"`
	ProjectID string      `json:"projectId"`
	TaskID    string      `json:"taskId"`
	Stage     ReviewStage `json:"stage"`
	Body      string      `json:"body"`
	CreatedAt time.Time   `json:"createdAt"`
	RunID     string      `json:"runId,omitempty"`
}

type AgentState struct {
	Workflows      []AgentWorkflow `json:"workflows"`
	Runs           []AgentRun      `json:"runs"`
	ReviewComments []ReviewComment `json:"reviewComments"`
}

type AgentTaskSnapshot struct {
	Workflow       AgentWorkflow   `json:"workflow"`
	ChatSessionID  string          `json:"chatSessionId,omitempty"`
	Runs           []AgentRun      `json:"runs"`
	ReviewComments []ReviewComment `json:"reviewComments"`
	DirtyFiles     []string        `json:"dirtyFiles"`
	AdapterState   string          `json:"adapterState"`
	Resumable      bool            `json:"resumable"`
	Interrupted    bool            `json:"interrupted"`
}
